gettingstarted/index.md

```diff
@@ -2,7 +2,462 @@
+
+You can think of a Rest Action as being a function `RestContext =>
+RestResponse[T]`.
+
+`RestContext` provides context about the request, such as the (parsed) request
+body, authentication information, fields of the object requested, paging and the
+raw `play.api.mvc.Request`.[^2]
+
+`RestResponse[T]` is similar to an `Either`, and has two concrete
+implementations:
+
+ * `Ok[T]` which indicates processing was successful.
+ * `RestError` which indicates an error response, and contains an `NaptimeActionException`
+   instance containing details such as the http error code.
+
+Most `Ok` responses require some form of `Keyed(key, value)` return type.
+`Keyed` is just a handy wrapper class to pair a model with its key.
+
+#### Getting started with GET
+
+Typically, rest actions are built using the included `Rest` builder object
+included with `RestActionHelpers`.
+
+Assuming `UserStore` has a method `def get(id: Int): Option[User]`, then we can
+write our get Rest Action as such:
+
+```scala
+def get(id: Int) = Rest.get { context =>
+  OkIfPresent(Keyed(id, userStore.get(id)))
+}
+```
+
+API route: `GET /api/users.v1/:id`
+
+A `GET` RestAction expects a return type of `RestResponse[Keyed[Int, User]]`, since we're
+trying to obtain a User model.
+
+Naptime expects that the Get method argument is named `id`. The resource ID
+extracted from the last part of the URL will be passed into the method as the
+`id` parameter.
+
+`OkIfPresent` is a helper that transforms the result of `userStore.get(id)`:
+
+* `Some(Keyed(id, user))` becomes `Ok(Keyed(id, user))`, indicating we found the user requested.
+* `None` becomes `RestError` with a `404` http code, indicating requested resource was not found.
+
+#### Create
+
+Let's take a look a how we would enable creation of a user.
+
+```scala
+def create() = Rest
+  .jsonBody[User]()  // parse request body as User
+  .create { context =>
+
+    val user = context.body
+    val id = userStore.create(user)
+
+    // Could return Ok(Keyed(id, Some(user)))
+    // if we want the updated entity to be in the response
+    Ok(Keyed(id, None))
+  }
+```
+
+API route: `POST /api/users.v1`
+
+A Create RestAction expects a return type of `RestResponse[Keyed[Int, Option[User]]]`,
+where the `Int` is the id of the newly created object, and `Option[User]` is to
+optionally return the created user model in the response.
+
+Here we've also used `Rest.jsonBody` to parse the request body so that
+`context.body` is of type `User` rather than `AnyContent`.
+In case you want to specify a maximum body length, you can also use
+the syntax `Rest.jsonBody[User](BODY_MAX_LENGTH)`.
+
+#### Update
+
+Here's how you would do an update of a user. An update is considered to be a
+wholesale replacement of the entity.
+
+```scala
+import org.coursera.playcour.naptime.Errors
+
+/*             SNIP                 */
+
+def update(id: Int) = Rest
+  .jsonBody[User]()  // parse request body as User
```

Comment:
I'd recommend removing `()`; I'd like to remove the default arguments at some point and just provide two overloads (so there's no decision to make about whether to include `()`).